### ISSUE TITLE
Interface for a custom algorithm to play as a bot

### ISSUE THREAD

**Issue by stephen-lazarionok** [lightvector/KataGo#901]
Is there any described interface that I can leverage to plug in my own algorithm?

Also what's the purpose of cpp and python source codes?

**Comment by stephen-lazarionok** [comment]
I would like to create my own bot that will compete with the bots published.

I see two options available:

1. Create a bot from scratch and implement GTP protocol - https://www.lysator.liu.se/~gunnar/gtp/ as you mentioned
2. Train my own NN that has the same inputs / outputs (e.g. "NN interface") as NN defined in your codebase. In this case I can leverage KataGo "ecosystem". For instance emulates matches between different bots.

For option (2) can you please document the inputs/outputs? Do you think it's worth doing this?

**Comment by lightvector** [comment]
No, it's not worth doing this. You should implement your own bot. Take a look at the jtauber repo there - it already provides all the starting point you need for GTP. And if you're going to train your own neural net then enough will likely be different that you actively don't want to use KataGo - because it would only create extra work for you to support the wide variety of training targets, different rulesets, etc. 

A far better point to start would be to just download some pro games databases and use https://github.com/mattheww/sgfmill to read the files and follow a pytorch tutorial on how to do supervised learning to predict the next move with a convolutional neural net, then plug that into the GTP framework from the jtauber repo and go from there.

**Comment by stephen-lazarionok** [comment]
Understood. Thanks for the clarifications. Then the question is how can I make the games happening between my model and KataGo models using GTP protocol?

**Comment by lightvector** [comment]
As far as I know, the options aren't great for that because the existing options are *just* passable enough that nobody has bothered making better ones, and the KataGo repo does *not* contain a solution for this anyways. For KataGo people have generally relied on gogui-twogtp (search online for it, it's a old java-based program), and the other option is "gomill" which is a python 2 (NOT python 3) predecessor to sgfmill, which the author of that library only had the capacity to update the SGF handling parts to python 3 but not the GTP related parts. I have an old sandbox script attached that I modify as needed that uses gomill when I need more flexibility than gogui-twogtp allows.

[custom_gtp_runner.zip](https://github.com/lightvector/KataGo/files/14351294/custom_gtp_runner.zip)

**Comment by stephen-lazarionok** [comment]
Basically it means that if I want to create a bot to compete with KataGo bots then it's easier to use "the interface" of NN you described.

**Comment by stephen-lazarionok** [comment]
Also, do you think it's worth implementing such an adapter or interface within KataGo project? Thus if I have "some algorithm" that implements GTP protocol I can easily simulate multiple games with other bots?
